## 4.8 Grouping Data & Aggregating Variables

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#  creating a string of the path for .pkl file from main folder
path = r"C:/Users/Soni/7-7-23(Instacart Basket Analysis)"

In [3]:
path

'C:/Users/Soni/7-7-23(Instacart Basket Analysis)'

In [4]:
#Importing pickel file
df_ords_prods_merged_upd = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [5]:
#create the subset of dataFrame
df = df_ords_prods_merged_upd[:1000000]

In [6]:
# check the shape of dataFrame
df.shape

(1000000, 18)

In [7]:
df.head(10)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   
5   3367565        1             6          2                  7   
6    550135        1             7          1                  9   
7   3108588        1             8          1                 14   
8   2295261        1             9          1                 16   
9   2550362        1            10          4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        average orders  
1  Regularly busy      Least busy        average orders  
2  Regularly busy      Least busy           most orders  
3      Least busy      Least busy        average orders  
4      Least busy      Least busy           most orders  
5  Regularly busy  Regularly busy        average orders  
6  Regularly busy     Busiest day           most orders  
7  Regularly busy     Busiest day           most orders  
8  Regularly busy     Busiest day           most orders  
9      Least busy      Least busy        average orders

**Grouping data with pandas**

In [8]:
df.groupby('product_name')

In [9]:
# by itself, the groupby() function is a little worthless in terms of output.
# ** you should always use the groupby() function as part of a series of steps, namely, the following:**
# 1. Split the data into groups based on some criteria.
# 2. Apply a function to each group separately.
# 3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.
# 4. So far, you’ve only completed the first step—splitting the data into groups (with the group being the “product_name” column).
# 5. Now, let’s take a look into the second step, which will involve some aggregation!

**Aggregating Data with agg()**

In [10]:
# Aggregation functions create summaries of selected columns and store these summary values in new columns. 
# These summaries usually take the form of a descriptive statistic, for instance, a mean, median, maximum, or minimum

**Performing a Single Aggregation**

In [11]:
#Recall that “order_number” refers to the number of orders placed by a given user.) 
#    ***This will involve two of the steps***
# 1. Split the data into groups based on “department_id.”
# 2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

In [12]:
df.groupby('department_id').agg({'order_number': ['mean']})


order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

**the average number of orders per user for each department ID You can easily see**

**for instance, that the department with an ID of 4 (produce) has a mean of around 19**

In [13]:
# There are some aggregations that can be conducted without use of the agg() function. 
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

**Rather than including the column name within square brackets**

**as in the example above, you could also set the column name off with a dot**

In [14]:
#df.groupby('department_id').order_number.mean()

In [15]:
#** Results are same**
#but
# Visual Appearance:
# Default Role: Square brackets have no other role in Python beyond indexing. 
# In general, stick to brackets when it comes to indexing!

**Performing Multiple Aggregations**

In [16]:
# All it comes down to is adding more arguments to your code.

df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

**Aggregating Data with transform()**

In [17]:
# it’s time to create flags for that data. 
# you’ll be creating a loyalty flag column in your ords_prods_merge dataframe
# Now, let’s map this task onto the three-step process introduced earlier:

#Split the data into groups based on the “user_id” column.
#Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
#Create a new column, “max_order,” into which you’ll place the results of your aggregation.

In [18]:
df_ords_prods_merged_upd['max_order'] = df_ords_prods_merged_upd.groupby(['user_id'])['order_number'].transform(np.max)

In [19]:
#After running this code:
# First, a new column called “max_order” is created,
#which will be what stores the maximum order number for each user (step 3).
#  Then, the ords_prods_merge dataframe is grouped by the “user_id” column (step 1).
# And finally, the transform() function is applied on the “order_number” column with the np.max argument (step 2).
# But what is this np? This is actually the NumPy library! 
# The max() function is a function included within NumPy that finds the max value within a column. 

In [20]:
# Now run this code & check the result with head()
df_ords_prods_merged_upd.head(20)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   
15   2293453       19             2          5                 14   
16   1973799       19             5          6                 12   
17    532817       19             7          4                 17   
18   1573906       21            10          3                 10   
19   1593000       31            10          3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77  

In [21]:
# each discrete value within the “max_order” column will appear in the dataframe 
#the same number of times as its value correspond to user_id
#  By comparing the number of orders with the order number and 
#the max order value, you can ascertain the accuracy of your aggregation.

**Deriving Columns with loc()**

In [22]:
# With your new column ready to go, all that’s left is to create a flag that assigns
#a “loyalty” label to a user ID based on its corresponding max order value

In [23]:
df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [24]:
df_ords_prods_merged_upd.loc[(df_ords_prods_merged_upd['max_order'] <= 40) & (df_ords_prods_merged_upd['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [25]:
df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [26]:
df_ords_prods_merged_upd['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [27]:
# Check the value and new column with loyalty_flag
df_ords_prods_merged_upd.head(20)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   
15   2293453       19             2          5                 14   
16   1973799       19             5          6                 12   
17    532817       19             7          4                 17   
18   1573906       21            10          3                 10   
19   1593000       31            10          3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77  

In [28]:
# check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column

df_ords_prods_merged_upd[['user_id', 'max_order', 'loyalty_flag']].sample(30)

user_id  max_order      loyalty_flag
9886257     56191         46    Loyal customer
19301058    61943          9      New customer
14541402    42366         13  Regular customer
2482528    191355         36  Regular customer
15280589   101816         17  Regular customer
1685166     68142         21  Regular customer
4506947    161098         15  Regular customer
12852974   178556          8      New customer
29824888   109257         13  Regular customer
11854291   100612         96    Loyal customer
22711955    35315         31  Regular customer
29138536   172141          5      New customer
432795     200427         57    Loyal customer
11188930   198218         23  Regular customer
1459790     16680         99    Loyal customer
24400630   193310         13  Regular customer
28957224   192842         10      New customer
30266602     1432         20  Regular customer
16703784   199349         68    Loyal customer
20672316    38510         20  Regular customer
18871363    15936         90    Loyal customer
404996     185338         41    Loyal customer
845983     178918         36  Regular customer
32336258   175946          4      New customer
15935230    42765         32  Regular customer
6402066     61796         44    Loyal customer
17342816   138530         35  Regular customer
16109001    43034         33  Regular customer
29726882   143782          3      New customer
10054060   164694         36  Regular customer

## Exercise Task 2

**find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe.**
**Now, repeat this process for the entire dataframe.**

In [29]:
df_ords_prods_merged_upd.groupby('department_id').agg({'order_number': ['mean']})


order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Task 3

**Analyze the result. How do the results for the entire dataframe differ from those of the subset?**

In [30]:
df.groupby('department_id').agg({'order_number': ['mean']})


order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

**Yes there is lot of difference. In subset and whole Dataframe each user_id has almost 1000000 difference but subset user_id 17 has more value difference** 

## Task 4: Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [31]:
# This is done already with this result.....
#Regular customer    15876776
#Loyal customer      10284093
#New customer         6243990
#Name: loyalty_flag, dtype: int64

## Task 5: whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [32]:
#Recall that “prices” refers to the amount paid for orders by a given user.) 
#    ***This will involve two of the steps***
# 1. Split the data into groups based on “loyalty_flag.”
# 2. Apply the agg() function to each group to obtain the mean values for the “prices” column.

**Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer).**

**Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.**

In [33]:
df_ords_prods_merged_upd.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [34]:
# With Statistics calculation we can see max prices paid by customers is 99999.0 which is shocking.
# Prices of products purchased by New customer is more as compared to Loyal and Regular customer.
# Loyal Customer purchased price is less than new customers.

In [35]:
# lets check descriptive statistics for prices column.
df_ords_prods_merged_upd['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

**With this we can see maximum price is 99.99 and mean value is 11.98**

**so we can check the other abnormal values also**

In [36]:
# This will give us all duplicates in descending order
df_ords_prods_merged_upd['prices'].drop_duplicates().sort_values(ascending = False)

29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

**By using this function we have found very strange price 99999.0 and 14900**

**Lets check which product has this value**

In [37]:
strange_price = df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['prices'].isin([99999.0, 14900.0])]

In [38]:
strange_price.sample(20)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
10031398   2182838    54327             6          0                  9   
10033818   1905459   164421             6          3                 14   
10033947   2736323   169314             6          0                 15   
10032000   1156090    80550             2          0                 12   
29165895   3295643   119625             6          4                 13   
10030545    951636     9288             1          0                 11   
10031316   2567730    50877             4          6                 12   
10032019   1634963    82185            37          1                 15   
10033482    783982   146986            40          2                 16   
10034435    264674   192552             6          0                 12   
10032135    322711    89777             7          2                 15   
29165800    181742   104127            18          1                 15   
10032572   2923158   103925             2          0                  6   
10031360   1276526    52480             4          0                 18   
10033944   3411439   169258             9          0                  8   
10032635   2655496   107073             7          5                 13   
10034196   2359941   180006            13          4                 20   
10034212    629790   180844             6          6                 14   
10032586     53649   104622             1          5                 20   
10034613   2047868   198803            10          5                 13   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
10031398                    30.0       21553                  2          1   
10033818                     3.0       21553                  6          1   
10033947                    30.0       21553                 12          1   
10032000                     7.0       21553                  1          1   
29165895                    21.0       33664                  3          0   
10030545                     NaN       21553                  2          0   
10031316                     2.0       21553                  2          1   
10032019                     5.0       21553                  8          1   
10033482                     7.0       21553                  3          1   
10034435                    10.0       21553                  6          1   
10032135                    30.0       21553                 15          1   
29165800                    14.0       33664                  6          1   
10032572                    20.0       21553                 14          0   
10031360                    21.0       21553                  4          1   
10033944                    23.0       21553                  8          1   
10032635                     6.0       21553                  7          1   
10034196                    10.0       21553                  8          1   
10034212                    18.0       21553                  4          1   
10032586                     NaN       21553                  2          0   
10034613                    10.0       21553                  8          0   

                              product_name  aisle_id  department_id   prices  \
10031398  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10033818  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10033947  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10032000  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
29165895             2 % Reduced Fat  Milk        84             16  99999.0   
10030545  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10031316  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10032019  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10033482  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.

*As we see in this sample---this strange price belongs to (Lowfat 2% Milkfat Cottage Cheese) & (2 % Reduced Fat Milk)*

*These products could not have these price range so we need to correct in the dataframe*

**Now we will check this strange price belongs to which product_id**

In [39]:
strange_price.describe()

order_id        user_id  order_number    order_dow  \
count  5.127000e+03    5127.000000   5127.000000  5127.000000   
mean   1.708876e+06  106117.185684     15.302126     2.750536   
std    9.838956e+05   58594.982659     16.691849     2.010892   
min    4.300000e+01      17.000000      1.000000     0.000000   
25%    8.642320e+05   58136.000000      4.000000     1.000000   
50%    1.693799e+06  106021.000000      9.000000     3.000000   
75%    2.567244e+06  156389.000000     20.000000     5.000000   
max    3.421047e+06  206049.000000     99.000000     6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        5127.000000  5127.000000  5127.000000         5127.0   
mean            7.986347     0.613614   104.732592           16.0   
std             6.868889     0.486968     8.231346            0.0   
min             1.000000     0.000000    84.000000           16.0   
25%             3.000000     0.000000   108.000000           16.0   
50%             6.000000     1.000000   108.000000           16.0   
75%            11.000000     1.000000   108.000000           16.0   
max            72.000000     1.000000   108.000000           16.0   

             prices    max_order  
count   5127.000000  5127.000000  
mean   26485.547494    29.791691  
std    29186.636448    24.929207  
min    14900.000000     3.000000  
25%    14900.000000    12.000000  
50%    14900.000000    21.000000  
75%    14900.000000    41.000000  
max    99999.000000    99.000000

**so these product_id belongs to department_id = 16. we need to check it detail**

In [40]:
#which products belongs to this product_id
df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['product_id'].isin([21553, 33664])]

order_id  user_id  order_number  order_dow  order_hour_of_day  \
10030345    912404       17            12          2                 14   
10030346    603376       17            22          6                 16   
10030347   3264360      135             2          2                 21   
10030348    892534      135             3          0                  8   
10030349    229704      342             8          1                 19   
...            ...      ...           ...        ...                ...   
29166209   2249946   204099            29          0                  8   
29166210   2363282   204099            31          0                  9   
29166211   3181945   204395            13          3                 15   
29166212   2486215   205227             7          3                 20   
29166213   2232716   205227            10          0                 11   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
10030345                     5.0       21553                  5          0   
10030346                     4.0       21553                  3          1   
10030347                    13.0       21553                  6          0   
10030348                    12.0       21553                  3          1   
10030349                    30.0       21553                  9          0   
...                          ...         ...                ...        ...   
29166209                     4.0       33664                  1          0   
29166210                     2.0       33664                  1          1   
29166211                     8.0       33664                 25          0   
29166212                     4.0       33664                  8          0   
29166213                     4.0       33664                 23          1   

                              product_name  aisle_id  department_id   prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030346  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030347  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030348  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030349  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
...                                    ...       ...            ...      ...   
29166209             2 % Reduced Fat  Milk        84             16  99999.0   
29166210             2 % Reduced Fat  Milk        84             16  99999.0   
29166211             2 % Reduced Fat  Milk        84             16  99999.0   
29166212             2 % Reduced Fat  Milk        84             16  99999.0   
29166213             2 % Reduced Fat  Milk        84             16  99999.0   

         _merge     price_range_loc     busiest_day    busiest_days  \
10030345   both  High-range product  Regularly busy  Regularly busy   
10030346   both  High-range product  Regularly busy  Regularly busy   
10030347   both  High-range product  Regularly busy  Regularly busy   
10030348   both  High-range product     Busiest day     Busiest day   
10030349   both  High-range product  Regularly busy     Busiest day   
...         ...                 ...             ...             ...   
29166209   both  High-range product     Busiest day     Busiest day   
29166210   both  High-range product     Busiest day     Busiest day   
29166211   both  High-range product  Regularly busy      Least busy   
29166212   both  High-range product  Regularly busy      Least busy   
29166213   both  High-range product     Busiest day     Busiest day   

         busiest_period_of_day  max_order      loyalty_flag  
10030345           most orders         40  Regular customer  
10030346           most orders         40  Regular customer  
10030347        average orders          4      New customer  
10030348        average orders          4      New customer  
10030349        average orders         16  Regul

*these product_ids belongs to Lowfat 2% Milkfat Cottage Cheese & 2 % Reduced Fat Milk*

In [41]:
# Create a filter 'milk_product' for records with the department_id value of 16, 
# and then 'dairy_without_weird' to exclude records (without very high prices)
# using ~ to exclude the data

milk_product = df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['department_id'].isin([16])]
milk_product_withoutstrange_price = milk_product.loc[~milk_product['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
milk_product_withoutstrange_price.sample(20)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id  order_number  order_dow  order_hour_of_day  \
5475149    1207792   151083            38          0                 10   
22695822   1834979   161589            12          6                 10   
6330384    1922203    45781            34          5                 16   
10330846    415708    73131            13          4                 14   
3888043     755138   202714            17          1                 20   
9066051     977363    22299            62          1                 10   
18025174    278370    93446             2          3                 11   
8048162    1877729   132605             1          5                 11   
19063070   2651943    56459            12          6                  9   
28997155   1581918    46782            10          5                 11   
11007392   2129219    94401            54          2                  8   
8435314     828438   191114            26          4                  8   
11839794   1506327    14494            18          6                  8   
8010913     831844    61822             6          0                 16   
15228363   1817939    53623             1          5                 16   
1411772    1935263    75825            83          3                 11   
26431843   1024837    83330            25          6                 12   
28073781   1049940   186938             6          4                 20   
8190373      15238   198180             9          0                 14   
24565585   1904562   155862            21          6                 16   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
5475149                      7.0       42828                 11          1   
22695822                    19.0       17993                  2          1   
6330384                      4.0       42736                  3          1   
10330846                     6.0       20156                 19          0   
3888043                     14.0       22035                 20          1   
9066051                      4.0       27086                  1          1   
18025174                     8.0       37029                  3          1   
8048162                      NaN        8309                  9          0   
19063070                    22.0       17902                  1          1   
28997155                    12.0       29846                 12          0   
11007392                    20.0       27845                 17          1   
8435314                      7.0       44987                  7          1   
11839794                    14.0       45446                  6          1   
8010913                      7.0       16185                  4          1   
15228363                     NaN       38154                  1          0   
1411772                      4.0        7781                  4          1   
26431843                     5.0       40183                  1          1   
28073781                    25.0       18224                  3          1   
8190373                     17.0       18465                  3          1   
24565585                     7.0        9038                 16          1   

                                               product_name  aisle_id  \
5475149                                Whipped Cream Cheese       108   
22695822             Shredded Colby & Monterey Jack Cheeses        21   
6330384                                     Unsalted Butter        36   
10330846                                 Mixed Berry Yogurt       120   
3888043                         Organic Whole String Cheese        21   
9066051                                         Half & Half        53   
18025174                                Cream Cheese Spread       108   
8048162            Nonfat Icelandic Style Strawberry Yogurt       120   
19063070                                  Liquid Egg Whites        86   
28997155                     Unsalted European Style Butter      

In [42]:
#check maximum price for this department_id without strange price
milk_product_withoutstrange_price['prices'].max()

15.0

In [43]:
# Lets check for 2 % Reduced Fat Milk (with product_id 33664)
#Lets divide this product name in three parts and search any products contains this value comes in table:
# Create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird

reduced_fat_milk = milk_product_withoutstrange_price.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')

In [44]:
print('Sample output from reduced_fat_milk:')
reduced_fat_milk.sample(10)

Sample output from reduced_fat_milk:


order_id  user_id  order_number  order_dow  order_hour_of_day  \
12854456    917119    11050             7          0                 11   
20793060    769085   182381            12          2                 15   
11702506   1280193   113277            16          3                 14   
16337722   1665200    37635             5          5                 19   
20161302   2245566    10302            33          4                  3   
14571111   1595037   204975             5          0                 13   
14564366   3325896   169639            78          0                 19   
11707330   2858803   170297            13          4                 15   
12870276   1427142   151888             4          2                 10   
16013678    728842    76705            11          2                 14   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
12854456                    30.0        1940                  2          0   
20793060                     9.0       19820                  7          0   
11702506                     3.0       32478                  5          1   
16337722                     6.0       40174                 11          1   
20161302                     5.0       13166                  2          1   
14571111                     9.0       23909                  2          1   
14564366                     8.0       23909                  1          1   
11707330                    14.0       32478                  6          1   
12870276                    19.0        1940                  2          0   
16013678                    15.0        5785                 11          1   

                                  product_name  aisle_id  department_id  \
12854456           Organic 2% Reduced Fat Milk        84             16   
20793060         Grassmilk 2% Reduced Fat Milk        84             16   
11702506                   Reduced Fat 2% Milk        84             16   
16337722           2% Reduced Fat Organic Milk        84             16   
20161302  Organic Milk Reduced Fat, 2% Milkfat        84             16   
14571111                   2% Reduced Fat Milk        84             16   
14564366                   2% Reduced Fat Milk        84             16   
11707330                   Reduced Fat 2% Milk        84             16   
12870276           Organic 2% Reduced Fat Milk        84             16   
16013678           Organic Reduced Fat 2% Milk        84             16   

          prices _merge    price_range_loc     busiest_day    busiest_days  \
12854456     9.1   both  Mid-range product     Busiest day     Busiest day   
20793060    12.6   both  Mid-range product  Regularly busy  Regularly busy   
11702506    13.9   both  Mid-range product  Regularly busy      Least busy   
16337722    11.4   both  Mid-range product  Regularly busy  Regularly busy   
20161302    12.8   both  Mid-range product      Least busy      Least busy   
14571111     9.2   both  Mid-range product     Busiest day     Busiest day   
14564366     9.2   both  Mid-range product     Busiest day     Busiest day   
11707330    13.9   both  Mid-range product      Least busy      Least busy   
12870276     9.1   both  Mid-range product  Regularly busy  Regularly busy   
16013678     7.9   both  Mid-range product  Regularly busy  Regularly busy   

         busiest_period_of_day  max_order      loyalty_flag  
12854456           most orders         14  Regular customer  
20793060           most orders         63    Loyal customer  
11702506           most orders         43    Loyal customer  
16337722        average orders         33  Regular customer  
20161302         fewest orders         38  Regular customer  
14571111           most orders          7      New customer  
14564366        average orders         99    Loyal customer  
11707330           most orders         30  Regular customer  
12870276           most orders         12  Regular customer  
16013678           m

In [45]:
# get description of this: reduced_fat_milk
reduced_fat_milk.describe()

order_id        user_id   order_number      order_dow  \
count  2.022380e+05  202238.000000  202238.000000  202238.000000   
mean   1.711414e+06  103040.384151      18.528684       2.748702   
std    9.876774e+05   59620.791055      18.288088       2.022770   
min    5.000000e+00      11.000000       1.000000       0.000000   
25%    8.586890e+05   51496.000000       5.000000       1.000000   
50%    1.706996e+06  102925.000000      12.000000       3.000000   
75%    2.566178e+06  154743.000000      26.000000       5.000000   
max    3.421036e+06  206209.000000      99.000000       6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      202238.000000           190556.000000  202238.000000   
mean           13.181835               10.538445   18768.761009   
std             4.193794                8.380235   13345.840890   
min             0.000000                0.000000    1940.000000   
25%            10.000000                5.000000    5785.000000   
50%            13.000000                7.000000   19820.000000   
75%            16.000000               14.000000   32478.000000   
max            23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      202238.000000  202238.000000  202238.000000       202238.0   
mean            5.423051       0.782692      83.716626           16.0   
std             5.397505       0.412415       6.511429            0.0   
min             1.000000       0.000000      21.000000           16.0   
25%             2.000000       1.000000      84.000000           16.0   
50%             4.000000       1.000000      84.000000           16.0   
75%             7.000000       1.000000      84.000000           16.0   
max           137.000000       1.000000      91.000000           16.0   

              prices      max_order  
count  202238.000000  202238.000000  
mean        9.086900      35.834769  
std         3.064869      26.069071  
min         1.200000       3.000000  
25%         7.900000      14.000000  
50%         9.200000      30.000000  
75%        11.400000      51.000000  
max        14.700000      99.000000

**Observation: Now the mean, min and max prices for products that contain 'Reduced Fat', '2' and 'Milk' in the name are: 9.08, 1.2, 14.7.**

**Since this result makes more sense, the price of product ID 33664, named '2 % Reduced Fat Milk' will be reduced from 99.999 to 9.99**

In [46]:
# Second: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird

cottage_cheese = milk_product_withoutstrange_price.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [47]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)

Sample output from cottage_cheese:


order_id  user_id  order_number  order_dow  order_hour_of_day  \
16191970   2167187   128800            66          2                 16   
16192388   2782586   167998            11          4                 17   
16192197    681046   148244             7          3                 17   
29899346   1923388   197976             7          0                 11   
16190997    839455    50656             1          2                 16   
30576121   1072067    60655             5          2                 23   
16192177    545549   146073            13          6                 13   
29899108    855633   116984            24          2                 13   
16190506   3386865     8420            41          5                  9   
16191342   2550284    85410             6          2                 13   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
16191970                     5.0       27790                  5          0   
16192388                     3.0       27790                  4          1   
16192197                     1.0       27790                 10          1   
29899346                     7.0        6852                  5          0   
16190997                     NaN       27790                 12          0   
30576121                     3.0       11203                  6          0   
16192177                     7.0       27790                  1          0   
29899108                     6.0        6852                  8          1   
16190506                     3.0       27790                  3          1   
16191342                    13.0       27790                  2          1   

                                               product_name  aisle_id  \
16191970        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16192388        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16192197        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29899346         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
16190997        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
30576121  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...       108   
16192177        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29899108         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
16190506        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16191342        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   

          department_id  prices _merge    price_range_loc     busiest_day  \
16191970             16     7.7   both  Mid-range product  Regularly busy   
16192388             16     7.7   both  Mid-range product      Least busy   
16192197             16     7.7   both  Mid-range product  Regularly busy   
29899346             16     8.6   both  Mid-range product     Busiest day   
16190997             16     7.7   both  Mid-range product  Regularly busy   
30576121             16     2.4   both  Low-range product  Regularly busy   
16192177             16     7.7   both  Mid-range product  Regularly busy   
29899108             16     8.6   both  Mid-range product  Regularly busy   
16190506             16     7.7   both  Mid-range product  Regularly busy   
16191342             16     7.7   both  Mid-range product  Regularly busy   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
16191970  Regularly busy           most orders         99    Loyal customer  
16192388      Least busy        average orders         58    Loyal customer  
16192197      Least busy        average orders          7      New customer  
29899346     Busiest day           most orders         13  Regular customer  
16190997  Regularly busy           most orders          7      New customer  
30576121  Regularly busy         fewest orders          5      New customer  
16192177  Regularly busy           most orders         15  Regular customer  
29899108  Regularly busy           most orders   

In [48]:
# Quick statistics in cottage_cheese

print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number    order_dow  \
count  3.487000e+03    3487.000000   3487.000000  3487.000000   
mean   1.684106e+06  103778.688271     15.484084     2.849728   
std    9.785618e+05   58583.418992     16.808776     2.031540   
min    2.301000e+03      52.000000      1.000000     0.000000   
25%    8.438150e+05   55839.000000      4.000000     1.000000   
50%    1.673171e+06  107636.000000     10.000000     3.000000   
75%    2.535750e+06  151541.500000     20.000000     5.000000   
max    3.419042e+06  206072.000000     98.000000     6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        3487.000000             3218.000000   3487.000000   
mean           13.257241               12.898073  22348.791225   
std             4.297401                9.231128   8826.332190   
min             0.000000                0.000000    195.000000   
25%            10.000000                6.000000  11203.000000   
50%            13.000000               10.000000  27790.000000   
75%            16.000000               19.000000  27790.000000   
max            23.000000               30.000000  31030.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        3487.000000  3487.000000  3487.000000         3487.0   
mean            5.840551     0.697448   107.650703           16.0   
std             5.955836     0.459429     5.502322            0.0   
min             1.000000     0.000000    21.000000           16.0   
25%             2.000000     0.000000   108.000000           16.0   
50%             4.000000     1.000000   108.000000           16.0   
75%             8.000000     1.000000   108.000000           16.0   
max            57.000000     1.000000   108.000000           16.0   

            prices    max_order  
count  3487.000000  3487.000000  
mean      7.228219    30.531116  
std       1.757900    25.413902  
min       2.400000     3.000000  
25%       7.700000    11.000000  
50%       7.700000    22.000000  
75%       7.700000    43.000000  
max      11.800000    99.000000

**Observation: Now the mean, min and max prices for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name are: 7.23, 2.4, 11.8.**
**Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese' will be replaced from 14.900 and inputed by the mean value mentioned above**

In [49]:
# Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

df_ords_prods_merged_upd['prices'] = df_ords_prods_merged_upd['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [50]:
# Check the statistics of the new 'prices' column in df_orders_products_merged

print('The statistics of the new prices column in df_orders_products_merged:')
df_ords_prods_merged_upd['prices'].describe()

The statistics of the new prices column in df_orders_products_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

**Observation: Now the mean, std, min and max results make more sense than Output [30]**

In [51]:
# Basic statistics of df_orders_products_merged by 'loyalty_flag' with the updated 'prices' column

print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
df_ords_prods_merged_upd.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

**Observation: Although the result remains the same As in starting Step 5:**
**new customer holds the highest purchase price followed by regular customer then loyal customer,** 
**now the mean value becomes lower and max value is more reasonable (instead of 99999, now 25)**

## Task 6:  Create a spending flag for each user based on the average price across all their orders

*1.If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”*

*2.If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”*

**Aggregating Data with transform()**

In [52]:
# it’s time to create flags for that data. 
# you’ll be creating a loyalty flag column in your ords_prods_merge dataframe
# Now, let’s map this task onto the three-step process introduced earlier:

#Split the data into groups based on the “user_id” column.
#Apply the transform() function on the “prices” column to generate the maximum orders for each user.
#Create a new column, “item_price,” into which you’ll place the results of your aggregation.

In [53]:
#Always check transform(np.?.max/average/median/mean?
df_ords_prods_merged_upd['item_price'] = df_ords_prods_merged_upd.groupby(['user_id'])['prices'].transform(np.average)

In [54]:
# Now run this code & check the result with head()
df_ords_prods_merged_upd.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        average orders        10   
1  Regularly busy      Least busy        average orders        10   
2  Regularly busy      Least busy           most orders        10   
3      Least busy      Least busy        average orders        10   
4      Least busy      Least busy           most orders        10   

   loyalty_flag item_price  
0  New customer   6.367797  
1  New customer   6.367797  
2  New customer   6.367797  
3  New customer   6.367797  
4  New customer   6.367797  

[5 rows x 21 columns]

**Deriving Columns with loc()**

In [55]:
# With your new column ready to go, all that’s left is to create a flag that assigns
#a “loyalty” label to a user ID based on its corresponding item_price value

In [56]:
df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['item_price'] >= 10, 'spending_flag'] = 'High spender'

In [57]:
df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['item_price'] < 10, 'spending_flag'] = 'Low spender'

In [58]:
df_ords_prods_merged_upd['spending_flag'].value_counts(dropna = False)

Low spender     32285182
High spender      119677
Name: spending_flag, dtype: int64

In [59]:
df_ords_prods_merged_upd.head(20)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   
15   2293453       19             2          5                 14   
16   1973799       19             5          6                 12   
17    532817       19             7          4                 17   
18   1573906       21            10          3                 10   
19   1593000       31            10          3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0          Soda  ...     9.0    both  Mid-range product  Regularly busy   
1          Soda  ...     9.0    both  Mid-range product  Regularly busy   
2          Soda  ...     9.0    both  Mid-range product  Regularly busy   
3          Soda  ...     9.0    both  Mid-range product      Least busy   
4          Soda  ...     9.0    both  Mid-range product      Least busy   
5          Soda  ...     9.0    both  Mid-range product  Regularly busy   
6          Soda  ...     9.0    both  Mid-range product  Regularly busy   
7          Soda  ...     9.0    both  Mid-range product  Regularly busy   
8          Soda  ...     9.0    both  Mid-range product  Regularly busy   
9          Soda  ...     9.0    both  Mid-range product      Least busy   
10         Soda  ...     9.0    both  Mid-range product  Regularly busy   
11         Soda  ...     9.0    both  Mid-range product  Regularly busy   
12         Soda  ...     9.0    both  Mid-range product  Regularly

In [60]:
# check the accuracy of the 'loyalty_flag' column with output from columns 'user_id','item_price', and 'loyalty_flag' column
df_ords_prods_merged_upd[['user_id', 'item_price', 'loyalty_flag']].sample(30)

user_id  item_price      loyalty_flag
14960661    10521    6.327826    Loyal customer
8412011    195724    7.053226      New customer
21375390   155919    7.389947  Regular customer
971125      79482    7.798964  Regular customer
8118938     74109    7.361702  Regular customer
29088679    98193    8.147222  Regular customer
3883275    186599    8.063800    Loyal customer
13073621   181293    8.034127  Regular customer
68890        2663    7.323851    Loyal customer
9697558     39526    8.706863  Regular customer
10450602    10179    7.045600  Regular customer
7512272    200327    8.083996    Loyal customer
25983230    88223    7.459184  Regular customer
17861293   139091    7.925926      New customer
17932249   185811    7.742211  Regular customer
11143543   111798    8.095313      New customer
19690411    76345    7.335548    Loyal customer
9987628    115744    8.609615  Regular customer
21244904    60132    7.497716  Regular customer
22789783   198832    7.486172    Loyal customer
4991394    172985    8.638349    Loyal customer
12460027   159747    7.874041  Regular customer
10710879    76295    7.569883    Loyal customer
22088712    47336    7.330293  Regular customer
19465990    76103    8.022879  Regular customer
28970063   201798    6.849658    Loyal customer
14356104    79433    9.962500      New customer
26797160    11360    6.581250      New customer
20720652   157340    7.910138      New customer
9590200    119382    7.571753    Loyal customer

In [61]:
df_ords_prods_merged_upd.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                      mean
loyalty_flag              
Loyal customer    7.773526
New customer      7.801190
Regular customer  7.798239

In [62]:
#Replace the wrong values with correct values:
#df_ords_prods_merged_upd.replace(to_replace = {'item_price': {14900.0 : 14.9}}, value = None)
#df['column name'] = df['column name'].replace(['old value'], 'new value')
#df_ords_prods_merged_upd['item_price '] = df_ords_prods_merged_upd['item_price'].replace(['14900.0'], '14.9')

In [63]:
# Check the frequency

df_ords_prods_merged_upd['spending_flag'].value_counts(dropna = False)

Low spender     32285182
High spender      119677
Name: spending_flag, dtype: int64

## Task 7:

## Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 

**1.If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”**

**2.If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”**

**3.If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”**

*Aggregating Data with transform()*

In [64]:
df_ords_prods_merged_upd['median_prior_orders'] = df_ords_prods_merged_upd.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [65]:
# Now run this code & check the result with head()
df_ords_prods_merged_upd.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc     busiest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag item_price  \
0  Regularly busy        average orders        10  New customer   6.367797   
1      Least busy        average orders        10  New customer   6.367797   
2      Least busy           most orders        10  New customer   6.367797   
3      Least busy        average orders        10  New customer   6.367797   
4      Least busy           most orders        10  New customer   6.367797   

   spending_flag median_prior_orders  
0    Low spender                20.5  
1    Low spender                20.5  
2    Low spender                20.5  
3    Low spender                20.5  
4    Low spender                20.5  

[5 rows x 23 columns]

In [66]:
# Creating a 'order_frequency_flag' flag

df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [67]:
# Creating a 'order_frequency_flag' flag

df_ords_prods_merged_upd.loc[(df_ords_prods_merged_upd['median_prior_orders'] > 10) & (df_ords_prods_merged_upd['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [68]:
# Creating a 'order_frequency_flag' flag

df_ords_prods_merged_upd.loc[df_ords_prods_merged_upd['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [69]:
df_ords_prods_merged_upd.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day    busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy      Least busy   
2         Soda  ...  Mid-range product  Regularly busy      Least busy   
3         Soda  ...  Mid-range product      Least busy      Least busy   
4         Soda  ...  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order  loyalty_flag item_price spending_flag  \
0        average orders        10  New customer   6.367797   Low spender   
1        average orders        10  New customer   6.367797   Low spender   
2           most orders        10  New customer   6.367797   Low spender   
3        average orders        10  New customer   6.367797   Low spender   
4           most orders        10  New customer   6.367797   Low spender   

   median_prior_orders   order_frequency_flag  
0                 20.5  Non-frequent customer  
1                 20.5  Non-frequent customer  
2                 20.5  Non-frequent customer  
3                 20.5  Non-frequent customer  
4                 20.5  Non-frequent customer  

[5 rows x 24 columns]

In [70]:
# Check the frequency

df_ords_prods_merged_upd['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [71]:
# Check the null values

print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
df_ords_prods_merged_upd[df_ords_prods_merged_upd['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

Records in df_orders_products_merged with null values in column order_frequency_flag:


user_id  days_since_prior_order  median_prior_orders  \
13645692   159838                     NaN                  NaN   
17251990   159838                     NaN                  NaN   
17622767   159838                     NaN                  NaN   
24138593   159838                     NaN                  NaN   
25880002   159838                     NaN                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

**This NaN values are only 5 so we can leave them**

## Task 8:
 ### Ensure your notebook is clean and structured and that your code is well commented.   

## Task 9:

**Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder**

In [72]:
df_ords_prods_merged_upd.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_grouping.pkl'))